### DISCUSSION TITLE
Async initialization of custom connector

### DISCUSSION THREAD

**Discussion by coolstrong** [algolia/instantsearch#6045] [Q&A] [answered]
I am building custom connector for a fairly specific task and I need to retrieve every possible facet value in advance for a widget to work (so before `getWidgetSearchParameters` and `getWidgetRenderState` methods). As far as i'm concerned, the only way to  retrieve all values is to search with empty query (e.g. `helper.searchForFacetValues(attribute, "", 100)`). But this is an async function, and `getWidgetSearchParameters` is synchronous, so there is no way to await all possible facet values.

Theoretically, this may be a job for an `init` method, but [the guide explicitly prohibits that](https://www.algolia.com/doc/guides/building-search-ui/widgets/create-your-own-widgets/js/#set-up-the-lifecycle). I tried nevertheless, and it results in UI endless loading. 

Currently I just fetch values in `refine` function if they are not present yet in connector state, but this way I cannot correctly implement `getWidgetSearchParameters` so search can be accurately restored from url.

**Comment by Haroenv** [comment, marked as answer]
This isn't possible, InstantSearch is designed to load synchronously to limit the number of requests happening and ensure a fast interface. Could you set `maxValuesPerFacet` to a high number instead and set `facets` or  `disjunctiveFacets` to read from the regular response?

**Reply by coolstrong** [reply]
Understood, I will try that, thanks!

_UPD:_ I'm using nextjs so I solved the issue by fetching facet values in a server component and passing it to my widget as prop.
